hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/files/StreamFileProducerBase.java

```diff
@@ -0,0 +1,355 @@
+     * @param currentSidecarFileWriters the list of sidecars that have been written so far, the last being the latest open one. Can
+     *                                  be null if there are no sidecar files yet.
+     * @param serializedSingleTransactionRecords List of serialized data from transactions
+     * @return the updated list of sidecar files written. Could be the same as passed in or have new ones added.
+     */
+    protected final List<SidecarFileWriter> handleSidecarItems(
+            @NonNull final Instant currentRecordFileFirstTransactionConsensusTimestamp,
+            @Nullable final List<SidecarFileWriter> currentSidecarFileWriters,
+            @NonNull final List<SerializedSingleTransactionRecord> serializedSingleTransactionRecords) {
+        // get or create the latest sidecar file writer
+        final List<SidecarFileWriter> sidecarFileWriters =
+                currentSidecarFileWriters != null ? new ArrayList<>(currentSidecarFileWriters) : new ArrayList<>();
+        SidecarFileWriter latestSidecarWriter =
+                sidecarFileWriters.isEmpty() ? null : sidecarFileWriters.get(sidecarFileWriters.size() - 1);
+        // iterate over the sidecar records, trying to write each one to sidecar file, creating new sidecar file if
+        // needed
+        try {
+            for (final SerializedSingleTransactionRecord serializedSingleTransactionRecord :
+                    serializedSingleTransactionRecords) {
+                final int numOfSidecarItems =
+                        serializedSingleTransactionRecord.sideCarItemsBytes().size();
+                for (int i = 0; i < numOfSidecarItems; i++) {
+                    // get the sidecar record
+                    final Bytes sidecarRecordBytes = serializedSingleTransactionRecord
+                            .sideCarItemsBytes()
+                            .get(i);
+                    // get the serialized bytes
+                    final TransactionSidecarRecord sideCarItem =
+                            serializedSingleTransactionRecord.sideCarItems().get(i);
+
+                    // get the kind of the sidecar record
+                    final var kind = sideCarItem.sidecarRecords().kind();
+                    // check if we have a latestSidecarWriter, if not create one
+                    if (latestSidecarWriter == null) {
+                        // create a new writer
+                        latestSidecarWriter = new SidecarFileWriter(
+                                getSidecarFilePath(
+                                        currentRecordFileFirstTransactionConsensusTimestamp,
+                                        sidecarFileWriters.size() + 1),
+                                compressFilesOnCreation,
+                                maxSideCarSizeInBytes);
+                        // add writer to list
+                        sidecarFileWriters.add(latestSidecarWriter);
```

Comment:
This code is duplicated twice in this method. It would be nice to factor out into another method, but I didn't see an easy way to do it that contributed to greater clarity.